### PR TITLE
Removing blob storage SDK leakage into the code base

### PR DESCRIPTION
**Pull request by agr**
Addresses: https://github.com/NuGet/Engineering/issues/5456

We have wrappers around blob storage SDK (`CloudBlobWrapper`, `CloudBlobContainerWrapper`, `CloudBlobClientWrapper`), unfortunately, they leak underlying SDK types as function arguments, return types and exceptions thrown, which defeats the purpose and complicates swapping the underlying implementation.

This change aims to stop that leakage. Introduced more wrappers around concepts that we use (`BlobListContinuationToken`, `CloudBlobCopyState`, `CloudBlobCopyStatus`, `CloudBlobLocationMode`, `CloudBlobPropertiesWrapper`, `ListingDetails`), propagated other wrappers that we had, but didn't use everywhere (`AccessConditionWrapper`, `FileUriPermissions`), wrapped `StorageException` with our own exception hierarchy following the checks existing in the code, simplifying catch expressions.

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <NuGetClientPackageVersion>6.9.1</NuGetClientPackageVersion>
     <ServerCommonPackageVersion>2.120.0</ServerCommonPackageVersion>
-    <NuGetJobsPackageVersion>4.3.0-main-9408418</NuGetJobsPackageVersion>
+    <NuGetJobsPackageVersion>4.3.0-agr-azst-9643015</NuGetJobsPackageVersion>
```
Comment (agr) [inline Directory.Build.props:5]: This would have to be updated with a separate PR as NuGet.Jobs release depends on this PR.

Comment (drewgillies) [reply Directory.Build.props:5]: This seems circular. Should we get away from this dependency?

File: src/NuGetGallery.Core/Services/CloudWrapperHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                return null;
+            }
+
+            return new AccessCondition { IfMatchETag = accessCondition.IfMatchETag, IfNoneMatchETag = accessCondition.IfNoneMatchETag };
+        }
+
+        public static SharedAccessBlobPermissions GetSdkSharedAccessPermissions(FileUriPermissions permissions)
+            => (SharedAccessBlobPermissions)permissions;
+
+        public static async Task<TResult> WrapStorageException<TResult>(Func<Task<TResult>> @delegate)
```
Comment (drewgillies) [inline src/NuGetGallery.Core/Services/CloudWrapperHelpers.cs:67]: Great pattern. +1

**Review by drewgillies — DISMISSED** [review]
Excellent! :)

File: src/NuGetGallery.Core/Services/CloudWrapperHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        {
+            try
+            {
+                return await @delegate();
+            }
+            catch (StorageException ex) when (ex.RequestInformation?.ExtendedErrorInformation?.ErrorCode == BlobErrorCodeStrings.ContainerNotFound)
```
Comment (lyndaidaii) [inline src/NuGetGallery.Core/Services/CloudWrapperHelpers.cs:73]: To avoid code duplication, consider creating a single method to handle the exceptions, and call this method from both WrapStorageException<TResult> and WrapStorageException.

**Review by lyndaidaii — APPROVED** [review]

File: src/NuGetGallery.Core/Services/CloudWrapperHelpers.cs (hunk 3 above)
Comment (agr) [reply src/NuGetGallery.Core/Services/CloudWrapperHelpers.cs:73]: Would you be able to suggest a change that would achieve it?

File: src/NuGetGallery.Core/Services/CloudBlobGenericNotFoundException.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace NuGetGallery
+{
+    public class CloudBlobGenericNotFoundException : CloudBlobStorageException
```
Comment (erdembayar) [inline src/NuGetGallery.Core/Services/CloudBlobGenericNotFoundException.cs:8]: Do we really need this generic one? I feel only `CloudBlobContainerNotFoundException` inherited from `CloudBlobStorageException` is enough.

Comment (agr) [reply src/NuGetGallery.Core/Services/CloudBlobGenericNotFoundException.cs:8]: We have existing code that checks specifically for "Container not found", "Blob not found" and "<something> not found" (i.e. request returned HTTP 404) cases. This setup supports all use cases without modifying code semantics.

File: Directory.Build.props (hunk 1 above)
Comment (erdembayar) [reply Directory.Build.props:5]: Sometimes this circular dependency makes hard to reason with runtime issue caused by dependency injection. Not sure exactly what was the actual dll consumed.

Comment (agr) [reply Directory.Build.props:5]: We have jobs in Gallery repo that use jobs base classes from NuGet.Jobs.

**Comment by erdembayar** [comment]
Could please add tracking issue into desc?

File: Directory.Build.props (hunk 1 above)
Comment (drewgillies) [reply Directory.Build.props]: I figured it would be something like that--I'll create a task for moving them, although that would also involve nugetizing some additional gallery dependencies to support them, I expect. DI, for one may make this a pipe dream.

**Comment by drewgillies** [comment]
> Could please add tracking issue into desc?
@erdembayar Done.

File: src/NuGetGallery.Core/Services/CloudBlobLocationMode.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGetGallery
+{
+    public enum CloudBlobLocationMode
```
Comment (erdembayar) [inline src/NuGetGallery.Core/Services/CloudBlobLocationMode.cs:6]: It looks this one is copy of https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.storage.retrypolicies.locationmode?view=azure-dotnet-legacy.
Should we add a comment above so the next person who is reading the code knows where those enum values are coming from?
